libs/grok/src/main/java/org/opensearch/grok/Grok.java

```diff
@@ -156,33 +161,85 @@ private void validatePatternBank() {
+        pathMap.put(initialPatternName, initialPath);
+        queue.push(new Frame(initialPatternName, initialPath, 0));
+
+        while (!queue.isEmpty()) {
+            Frame frame = queue.peek();
+            String patternName = frame.patternName;
+            Deque<String> path = frame.path;
+            int startIndex = frame.startIndex;
+            String pattern = patternBank.get(patternName);
+
+            if (visitedPatterns.contains(patternName)) {
+                queue.pop();
+                continue;
             }
-            int semanticNameIndex = pattern.indexOf(':', begin);
-            int end = syntaxEndIndex;
-            if (semanticNameIndex != -1) {
-                end = Math.min(syntaxEndIndex, semanticNameIndex);
+
+            visitedPatterns.add(patternName);
+            boolean foundDependency = false;
+
+            for (int i = startIndex; i < pattern.length(); i++) {
+                if (pattern.startsWith("%{", i)) {
+                    int begin = i + 2;
+                    int syntaxEndIndex = pattern.indexOf('}', begin);
+                    if (syntaxEndIndex == -1) {
+                        throw new IllegalArgumentException("Malformed pattern [" + patternName + "][" + pattern + "]");
+                    }
+
+                    int semanticNameIndex = pattern.indexOf(':', begin);
+                    int end = semanticNameIndex == -1 ? syntaxEndIndex : Math.min(syntaxEndIndex, semanticNameIndex);
+
+                    String dependsOnPattern = pattern.substring(begin, end);
+
+                    if (dependsOnPattern.equals(patternName)) {
+                        throwExceptionForCircularReference(patternName, pattern);
+                    }
+
+                    if (pathMap.containsKey(dependsOnPattern)) {
+                        path.removeFirst();
+                        throwExceptionForCircularReference(patternName, pattern, dependsOnPattern, path);
+                    }
+
+                    Deque<String> newPath = new ArrayDeque<>(path);
+                    newPath.push(dependsOnPattern);
+                    pathMap.put(dependsOnPattern, newPath);
+
+                    queue.push(new Frame(dependsOnPattern, newPath, 0));
+                    frame.startIndex = i + 1;
+                    foundDependency = true;
+                    break;
+                }
+            }
+
+            if (!foundDependency) {
+                pathMap.remove(patternName);
+                queue.pop();
+            }
+
+            if (queue.size() > MAX_PATTERN_DEPTH_SIZE) {
+                throw new IllegalArgumentException("Pattern references exceeded maximum depth of " + MAX_PATTERN_DEPTH_SIZE);
             }
-            String dependsOnPattern = pattern.substring(begin, end);
-            validatePatternBank(dependsOnPattern, path);
         }
-        path.pop();
+    }
+
+    private static class Frame {
+        String patternName;
+        Deque<String> path;
```

Comment:
Does this need to be a `Deque`? 

It feels like maybe a `List` would be sufficient. I'm not convinced that you need the stack/queue semantics.

Reply:
I guess you are correct, that as `ArrayList` implementation suits this better as these are limitations on operations related to path/newPath:
- Adding elements to the end of the list `newPath.add(dependsOnPattern)`.
- Copying the list to create newPath from path.
- Iterating through the list for operations such as checking dependencies and potentially removing the last element `path.remove(path.size() - 1)`. - this I think we can omit as well by taking care directly in `throwExceptionForCircularReference` instead of manipulating the path.